preprocessing/generate_from_api.py

```diff
@@ -0,0 +1,119 @@
+
+# pylint: disable=locally-disabled, line-too-long, broad-exception-caught, too-many-arguments
+
+class Preprocessor:
+    """Handles the preprocessing of ML data"""
+
+    def __init__(self: Self, stock_file_list: List[str], currency_file_list: List[str]) -> None:
+        self.cwd = os.getcwd() + '\\'
+        self.stock_files = stock_file_list
+        self.currency_files = currency_file_list
+        self.df = pd.DataFrame()
+
+    # end __init__
+
+    def rename_columns(self: Self, file_name: str, dataframe: pd.DataFrame):
+        """Applies a suffix to each column of a dataframe from it's
+        filename."""
+        suffix = file_name.partition('.')[0]
+        columns = dataframe.columns
+        new_columns = []
+
+        for col in columns:
+            if col != 'timestamp':
+                new_columns.append(col + '_' + suffix)
+            else:
+                new_columns.append('timestamp')
+
+        dataframe.columns = new_columns #PBR?
+
+    # end rename_columns
+
+    def concatenate(self: Self) -> pd.DataFrame:
+        """Takes the stock/currency files specifed
+        in the constructor and opens then concatenates
+        them into a single dataframe"""
+        stock_frames = []
+        currency_frames = []
+
+        # Convert CSV files into pandas dataframes
+        for cur_file in self.currency_files:
+            currency_frames.append(pd.read_csv(os.path.join(self.cwd + cur_file)))
+
+        for stock_file in self.stock_files:
+            stock_frames.append(pd.read_csv(os.path.join(self.cwd + stock_file)))
+
+
+        # Rename columns
+        for stock_f, stock_df in zip(self.stock_files, stock_frames):
+            self.rename_columns(stock_f, stock_df)
+
+        for currency_f, currency_df in zip(self.currency_files, currency_frames):
+            self.rename_columns(currency_f, currency_df)
+
+
+        # Concatenate all dataframes into one dataframe
+        all_frames = stock_frames + currency_frames
+
+        for frame in all_frames:
+            frame['timestamp'] = pd.to_datetime(frame['timestamp']).dt.date
+
+            # Catch first pass where df is empty
+            if self.df.empty:
+                self.df = frame
+            else:
+                self.df = self.df.merge(right=frame, on='timestamp')
+
+        return self.df
+
+        # end concatenate
+
+    def apply_preprocessing(self: Self, minmax=True, moving_average=True, moving_average_window=10, rsi_index=True) -> pd.DataFrame:
+        """Applies various preprocessing options to the 
+        final dataframe."""
+        # Convert timestamp to numerical value
+        self.df['timestamp'] = pd.to_numeric(pd.to_datetime(self.df['timestamp']))
+
+        if moving_average:
+            for col in self.df.columns:
+                if col != 'timestamp':
+                    self.df[col] = self.df[col].rolling(moving_average_window).mean()
+
+        if rsi_index:
+            pass #TODO
```

Comment:
When do we plan do to the RSI_index, and is it necessary for our project?

Reply:
RSI index measures the "momentum" of a stock, which basically tries to quantify if a stock is being overbought or oversold. I haven't included it yet because the calculation uses an averaging function similar to the "moving_average" function, and I'm not sure if that double-averaging will screw up the final result.

 I suspect that we won't know until we test it with just the moving average, just the RSI, and then with the moving average and the RSI index. Or, we could potentially add the moving average/RSI as another column to prevent over-averaging. 

But to answer your question, it isn't essential, but it will be interesting to test it to see if it can improve our model's results.

Reply:
I see, LGTM then